### PR TITLE
Fix PBSJobManager for qstat output with multiple lines

### PR DESCRIPTION
**Pull request by askask**

### CODE REVIEW

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy [hunk 1]
```diff
@@ -245,13 +245,13 @@ class PBSJobManager extends ClusterJobManager<PBSCommand> {
      * @param resultLines - Input of ExecutionResult object
      * @return map with jobid as key
      */
-    private Map<BEJobID, GenericJobInfo> processQstatOutput(List<String> resultLines) {
+    private Map<BEJobID, GenericJobInfo> processQstatOutput(String result) {
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy]: This is possible because you spit out XML now, right?

File: src/test/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManagerSpec.groovy [hunk 2]
```diff
@@ -87,11 +87,11 @@ class PBSJobManagerSpec extends Specification {
         def parms = JobManagerOptions.create().build()
         TestExecutionService testExecutionService = new TestExecutionService("test", "test")
         PBSJobManager jm = new PBSJobManager(testExecutionService, parms)
-        Method method = PBSJobManager.class.getDeclaredMethod("processQstatOutput", List)
+        Method method = PBSJobManager.class.getDeclaredMethod("processQstatOutput", String)
```
Comment (dankwart-de) [inline src/test/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManagerSpec.groovy]: Why do you use reflection here? Could you also make the method access public / package based?

File: src/test/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManagerSpec.groovy [hunk 3]
```diff
@@ -103,11 +103,11 @@ class PBSJobManagerSpec extends Specification {
...
```
Comment (dankwart-de) [inline src/test/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManagerSpec.groovy]: Dto. 
If you keep reflection, please pull 
Method method = PBSJobManager.class.getDeclaredMethod("processQstatOutput", String)
method.setAccessible 
method.invoke

to a function. It's code duplication.

Comment (dankwart-de) [reply src/test/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManagerSpec.groovy]: Please also think, if you can reduce code duplication in the tests. You could maybe store the job manager in class scope?

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy [hunk 4]
```diff
@@ -184,7 +184,7 @@ class PBSJobManager extends ClusterJobManager<PBSCommand> {
         }
 
         if (er != null && er.successful) {
-            queriedExtendedStates = this.processQstatOutput(er.resultLines)
+            queriedExtendedStates = this.processQstatOutput(er.resultLines.join("\n"))
         }else{
             logger.postAlwaysInfo("Extended job states couldn't be retrieved. \n Returned status code:${er.exitCode} \n result:${er.resultLines}")
```
[the remark this answers is not in the document]
Comment (askask) [reply src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy]: I removed logging and replaced it with exceptions everywhere

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy (hunk 1 above)
Comment (askask) [reply src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy]: Yes that's right

File: src/test/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManagerSpec.groovy (hunk 2 above)
Comment (askask) [reply src/test/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManagerSpec.groovy]: I removed the reflection

File: src/test/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManagerSpec.groovy (hunk 3 above)
Comment (askask) [reply src/test/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManagerSpec.groovy]: I removed the code duplication

File: src/main/groovy/de/dkfz/roddy/execution/RestExecutionService.groovy [hunk 5]
```diff
@@ -39,6 +38,8 @@ import org.apache.http.ssl.SSLContextBuilder
 import org.apache.http.ssl.SSLContexts
 import org.apache.http.ssl.TrustStrategy
 import org.apache.http.util.EntityUtils
+import org.slf4j.Logger
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/execution/RestExecutionService.groovy]: I am not really sure, why you use a different logger here. Beforehand we used the LoggerWrapper for everything. Do you want to change that?

File: src/main/groovy/de/dkfz/roddy/execution/RestExecutionService.groovy [hunk 6]
```diff
@@ -109,7 +109,7 @@ class RestExecutionService implements BEExecutionService {
         headers.add(new BasicHeader(HTTP.CONTENT_TYPE, "application/xml;charset=UTF-8"))
         headers.add(new BasicHeader("Accept", "application/xml"))
         RestResult result = execute(new RestCommand(RESOURCE_LOGON, body, headers, RestCommand.HttpMethod.HTTPPOST))
-        logger.severe("status code: " + result.statusCode)
+        logger.debug("response status: {} ", result.statusCode)
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/execution/RestExecutionService.groovy]: see above

File: src/main/groovy/de/dkfz/roddy/execution/jobs/BatchEuphoriaJobManager.groovy [hunk 7]
```diff
@@ -27,8 +26,6 @@ import java.util.concurrent.TimeoutException
 @CompileStatic
 abstract class BatchEuphoriaJobManager<C extends Command> {
 
-    private static final LoggerWrapper logger = LoggerWrapper.getLogger(BatchEuphoriaJobManager.class.getSimpleName())
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/BatchEuphoriaJobManager.groovy]: Why did you remove all the loggers?

File: src/main/groovy/de/dkfz/roddy/execution/jobs/BatchEuphoriaJobManager.groovy [hunk 8]
```diff
@@ -72,8 +69,7 @@ abstract class BatchEuphoriaJobManager<C extends Command> {
         this.queryOnlyStartedJobs = parms.trackOnlyStartedJobs
         this.userIDForQueries = parms.userIdForJobQueries
         if (!userIDForQueries && isTrackingOfUserJobsEnabled) {
-            logger.warning("Silently falling back to default job tracking behaviour. The user name was not set properly and the system cannot track the users jobs.")
-            isTrackingOfUserJobsEnabled = false
+            throw new BEException("The user name was not set properly and the system cannot track the users jobs.")
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/BatchEuphoriaJobManager.groovy]: At least for Roddy, this could come up as an issue, please open an issue describing this change.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/BatchEuphoriaJobManager.groovy [hunk 9]
```diff
@@ -218,7 +212,6 @@ abstract class BatchEuphoriaJobManager<C extends Command> {
     }
 
     int waitForJobsToFinish() {
-        logger.info("The user requested to wait for all jobs submitted by this process to finish.")
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/BatchEuphoriaJobManager.groovy]: info -> rare?

File: src/main/groovy/de/dkfz/roddy/execution/jobs/BatchEuphoriaJobManager.groovy [hunk 10]
```diff
@@ -338,18 +330,14 @@ abstract class BatchEuphoriaJobManager<C extends Command> {
...
-        try {
-            updateDaemonThread = Thread.startDaemon("Job state update daemon.", {
-                updateActiveJobList()
-                try {
-                    Thread.sleep(Math.max(cacheUpdateInterval.toMillis(), 10 * 1000))
-                } catch (InterruptedException e) {
-                    e.printStackTrace()
-                }
-            })
-        } catch (Exception ex) {
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/BatchEuphoriaJobManager.groovy]: try -> catch -> throw? 
Or try to catch known exceptions.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/ClusterJobManager.groovy [hunk 11]
```diff
@@ -22,7 +23,7 @@ import java.time.Duration
  */
 @CompileStatic
 abstract class ClusterJobManager<C extends Command> extends BatchEuphoriaJobManager<C> {
-    private static final LoggerWrapper logger = LoggerWrapper.getLogger(BatchEuphoriaJobManager.class.getSimpleName());
+    final static Logger logger = LoggerFactory.getLogger(ClusterJobManager.class)
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/ClusterJobManager.groovy]: Again, we'd have two different logger types in the same library.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/BatchEuphoriaJobManager.groovy (hunk 7 above)
Comment (askask) [reply src/main/groovy/de/dkfz/roddy/execution/jobs/BatchEuphoriaJobManager.groovy]: I just removed the logger that were unused

File: src/main/groovy/de/dkfz/roddy/execution/RestExecutionService.groovy (hunk 5 above)
Comment (askask) [reply src/main/groovy/de/dkfz/roddy/execution/RestExecutionService.groovy]: I think for a library the best approach is to use a logging façade, so users of the library are not required to use a specific logging framework.
See also this [answer](https://stackoverflow.com/a/9741118)
